### PR TITLE
Stop overall deployment if one of the deployment steps fails

### PR DESCRIPTION
**Pull request by fao89**

### CODE REVIEW

File: controllers/openstackdataplanenode_controller.go [hunk 1]
```diff
@@ -303,10 +303,12 @@ func (r *OpenStackDataPlaneNodeReconciler) GetInstanceRole(ctx context.Context,
...
-	err = r.Client.Get(ctx, client.ObjectKey{
-		Namespace: instance.Namespace,
-		Name:      instance.Spec.Role,
-	}, instanceRole)
+	if len(instance.Spec.Role) > 0 {
+		err = r.Client.Get(ctx, client.ObjectKey{
+			Namespace: instance.Namespace,
+			Name:      instance.Spec.Role,
+		}, instanceRole)
+	}
```
Comment (fultonj) [inline controllers/openstackdataplanenode_controller.go:311]: I agree about only running Get if there was a role set. Regardless of if we want to make the role mandatory or optional, either way we should have this code (306-311).  If we go optional, then we could have something like this.

  https://github.com/openstack-k8s-operators/dataplane-operator/pull/69

File: pkg/deployment/deployment.go [hunk 2]
```diff
@@ -364,13 +364,14 @@ func ConditionalDeploy(
...
 			log.Info(fmt.Sprintf("%s error", readyCondition))
-			err = fmt.Errorf("job: job.namespace %s job.name %s failed", ansibleEEJob.Namespace, ansibleEEJob.Name)
+			err = fmt.Errorf("failed: job.name %s job.namespace %s", ansibleEEJob.Name, ansibleEEJob.Namespace)
 			status.Conditions.Set(condition.FalseCondition(
 				readyCondition,
 				condition.ErrorReason,
 				condition.SeverityWarning,
 				readyErrorMessage,
 				err.Error()))
+			return ctrl.Result{}, err
```
Comment (fultonj) [inline pkg/deployment/deployment.go:374]: So if the job fails, then we'll return a failure. I agree.

**Review by fultonj — COMMENTED** [review]
+2
Let's fail early instead of giving the user the impression we're going to do something.

Since this includes failing early if the role is undefined we should consider if we want the role to be optional or required. If required, then we should update the API too so that the node CR can't be created unless it has a role. Alternatively, we can allow the role to be optional as per the following. 

https://github.com/openstack-k8s-operators/dataplane-operator/pull/69

Either way, this patch (68) and the above (69) are compatible.

File: controllers/openstackdataplanenode_controller.go (hunk 1 above)
Comment (slagle) [reply controllers/openstackdataplanenode_controller.go:311]: +1

**Review by slagle — APPROVED** [review]

**Pull request merged**